[docs] Adds qr.expo.dev reference doc (#28178)
expo/expo: docs/pages/more/qr-codes.mdx

---
title: qr.expo.dev
description: Reference for the QR code generator at qr.expo.dev.
---

import ImageSpotlight from '~/components/plugins/ImageSpotlight';

> qr.expo.dev is a cloud function that generates Expo-branded QR codes. This function creates QR codes for EAS Updates, which are used to preview updates in development builds and Expo Go.

Reviewer: ```suggestion
qr.expo.dev is a cloud function that generates Expo-branded QR codes. This function creates QR codes for [EAS Updates](/eas-update/introduction/), which are used to preview updates in [development builds](/develop/development-builds/introduction/) and Expo Go.
```

Let's add internal links here.

> For example, if you and your team have a development build, and you'd like to load the latest update on a specific build's channel, you could go to following endpoint to generate a QR code:

Reviewer: ```suggestion
For example, if you and your team have a development build, and you'd like to load the latest update on a specific build's channel, you could go to the following endpoint to generate a QR code:
```

```text
https://qr.expo.dev/eas-update?projectId=your-project-id&runtimeVersion=your-runtime-version&channel=your-channel
```

Which would produce the following QR code SVG:

<ImageSpotlight
  alt="QR code generated by qr.expo.dev"
  src="https://qr.expo.dev/eas-update?slug=your-slug&projectId=your-project-id&runtimeVersion=your-runtime-version&channel=your-channel"
  style={{ maxWidth: 200 }}
/>

This QR code represents the following URL:

```text
exp+your-slug://expo-development-client/?url=https://u.expo.dev/your-project-id?runtime-version=your-runtime-version&channel-name=your-channel
```

This URL will deep link into a development build and instruct it to fetch the latest update on the specified channel.

> **info** If sharing the URL is more convenient, you can request the URL directly by adding `format=url` to the query parameters.

## General

> The following parameters apply to the /eas-update endpoint.

Reviewer: ```suggestion
The following parameters apply to the `/eas-update` endpoint.
```

Since /eas-update is part of endpoint, let's use inline code syntax to highlight it.

### Base query parameters

> The following are base query parameters that can be included with any request to /eas-update.

Reviewer: ```suggestion
The following base query parameters can be included with any request to `/eas-update`.
```

Since /eas-update is part of endpoint, let's use inline code syntax to highlight it.

> | param                    | required | default    | description                                                                                          |

Reviewer: ```suggestion
| Param                    | Required | Default    | Description                                                                                          |
```

For consistency with the rest of our docs, let's capitalize each word in the header of the table.

| ------------------------ | -------- | ---------- | ---------------------------------------------------------------------------------------------------- |

> | `slug`                   | no       | exp        | To target Expo Go, use "exp". Otherwise, use `slug` from **app.json** to target a development build. |

Reviewer: ```suggestion
| `slug`                   | No       | exp        | Use [`slug`](/versions/latest/config/app/#slug) from [app config](/workflow/configuration/) to target a development build. Otherwise, use "exp" to target Expo Go. |
```

Let's make the description development-build first and internal links. Also, instead of mentioning **app.json**, let's use app config as a term since it represents different variants of app config file regardless of the extension (such as app.config.js/app.config.ts).

> | `appScheme` (deprecated) | no       | exp        | Replaced by `slug`. Use `slug` instead.                                                              |

Reviewer: ```suggestion
| `appScheme` (deprecated) | No       | exp        | Replaced by `slug`. Use `slug` instead.                                                              |
```

> | `host`                   | no       | u.expo.dev | The hostname used in the URL, must be a first-party URL.                                             |
> | `format`                 | no       | svg        | Endpoints respond with SVGs by default. To receive a plain text URL, use "url".                      |

Reviewer: ```suggestion
| `host`                   | No       | u.expo.dev | The hostname used in the URL. Must be a first-party URL.                                             |
| `format`                 | No       | svg        | Endpoints respond with SVGs by default. To receive a plain text URL, use `url`.                      |
```
- In the first description, what does a first-party URL means? Is it a common term that a reader should be aware of?
- Can we add inline syntax highlight for the "url" part to indicate it is part other QR code URL.

### Update by device traits

Preview and production builds make requests to the EAS Update service with `runtimeVersion` and `channel` properties. You can emulate this behavior with the following query parameters:

> | param            | required | description                      |
> | ---------------- | -------- | -------------------------------- |
> | `projectId`      | yes      | The ID of the project            |
> | `runtimeVersion` | yes      | The runtime version of the build |
> | `channel`        | yes      | The channel name of the build    |

Reviewer: ```suggestion
| Param            | Required | Description                      |
| ---------------- | -------- | -------------------------------- |
| `projectId`      | Yes      | The ID of the project.           |
| `runtimeVersion` | Yes      | The [runtime version](/eas-update/runtime-versions/) of the build. |
| `channel`        | Yes      | The channel name of the build.    |
```

#### Example

```text
https://qr.expo.dev/eas-update?projectId=your-project-id&runtimeVersion=your-runtime-version&channel=your-channel
```

### Update by ID

You can create a QR code for a specific update given its platform-specific ID.

> | param      | required | description          |
> | ---------- | -------- | -------------------- |
> | `updateId` | yes      | The ID of the update |

Reviewer: ```suggestion
| Param      | Required | Description          |
| ---------- | -------- | -------------------- |
| `updateId` | Yes      | The ID of the update. |
```

#### Example

```text
https://qr.expo.dev/eas-update?updateId=your-update-id
```

### Update by group ID

You can create a QR code for an update group given the update's group ID.

> | param       | required | description                |
> | ----------- | -------- | -------------------------- |
> | `projectId` | yes      | The ID of the project      |
> | `groupId`   | yes      | The ID of the update group |

Reviewer: ```suggestion
| Param       | Required | Description                |
| ----------- | -------- | -------------------------- |
| `projectId` | Yes      | The ID of the project.      |
| `groupId`   | Yes      | The ID of the update group. |
```

#### Example

```text
https://qr.expo.dev/eas-update?projectId=your-project-id&groupId=your-update-id
```

### Update by branch ID

You can create a QR code with a branch's ID, which will return the latest update available on that branch.

> | param       | required | description           |
> | ----------- | -------- | --------------------- |
> | `projectId` | yes      | The ID of the project |
> | `branchId`  | yes      | The ID of the branch  |

Reviewer: ```suggestion
| Param       | Required | Description           |
| ----------- | -------- | --------------------- |
| `projectId` | Yes      | The ID of the project. |
| `branchId`  | Yes      | The ID of the branch.  |
```

#### Example

```text
https://qr.expo.dev/eas-update?projectId=your-project-id&branchId=your-branch-id
```

### Update by channel ID

You can create a QR code with a channel's ID, which will return the latest update available on the branch or branches that are mapped to that channel.

> | param       | required | description           |
> | ----------- | -------- | --------------------- |
> | `projectId` | yes      | The ID of the project |
> | `channelId` | yes      | The ID of the channel |

Reviewer: ```suggestion
| Param       | Required | Description           |
| ----------- | -------- | --------------------- |
| `projectId` | Yes      | The ID of the project. |
| `channelId` | Yes      | The ID of the channel. |
```
